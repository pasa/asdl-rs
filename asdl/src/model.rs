#[derive(Debug)]
pub struct Asdl {
    pub types: Vec<Type>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn get_type_by_name(&self, name: &str) -> Option<&Type> {
        self.types.iter().find(|t| t.id() == name)
    }
}

#[derive(Debug)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl Type {
    fn id(&self) -> String {
        match self {
            Type::SumType(sty) => sty.id.clone(),
            Type::ProdType(pty) => pty.id.clone(),
        }
    }
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub arity: Arity,
}

#[derive(Debug)]
pub enum Arity {
    Optional,
    Required,
    Repeated,
}
