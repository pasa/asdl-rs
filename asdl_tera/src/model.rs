use std::collections::HashMap;
use serde::Serialize;
use asdl;
use super::Result;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: HashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn parse(asdl: &str) -> Result<Asdl> {
        Ok(Asdl::new(asdl::Asdl::parse(asdl)?))
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub is_prod_type: bool, //always false
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub is_prod_type: bool, //always true
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub is_required: bool,
    pub is_optional: bool,
    pub is_repeated: bool,
}
