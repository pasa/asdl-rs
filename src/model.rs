use std::fmt::Write;

use rustc_hash::FxHashMap;
use heck::MixedCase;
use linked_hash_set::LinkedHashSet;
use serde::Serialize;

use crate::parser;

#[derive(Serialize, Debug)]
pub struct Asdl {
    sum_types: Vec<SumType>,
    prod_types: Vec<ProdType>,
}

#[derive(Serialize, Debug)]
pub(crate) struct SumType {
    id: String,
    constructors: Vec<Constructor>,
    attributes: Vec<Field>,
}

#[derive(Serialize, Debug)]
pub(crate) struct Constructor {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct ProdType {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

impl Asdl {
    pub(crate) fn new(root: &parser::Root) -> Self {
        let mut res = Asdl { prod_types: vec![], sum_types: vec![] };
        for d in root.types() {
            match d.kind() {
                parser::TypeKind::SumType(t) => {
                    res.sum_types.push(sum_type(t));
                }
                parser::TypeKind::ProdType(t) => {
                    res.prod_types.push(prod_type(t));
                }
            }
        }
        res
    }
}

fn sum_type(node: &parser::SumType) -> SumType {
    let id = node.type_id().text().to_string();
    let constructors = node.constructors().map(constructor).collect();
    let attributes = if let Some(attrs) = node.attrs() {
        let mut names = FieldNames::default();
        attrs.fields().map(|f| field(f, &mut names)).collect()
    } else {
        vec![]
    };
    SumType { id, constructors, attributes }
}

fn constructor(node: &parser::Constr) -> Constructor {
    let id = node.id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    Constructor { id, fields }
}

fn field(node: &parser::Field, names: &mut FieldNames) -> Field {
    match node.kind() {
        parser::FieldKind::Single(f) => Field::single(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Opt(f) => Field::option(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Sequence(f) => Field::sequence(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
    }
}

fn prod_type(node: &parser::ProdType) -> ProdType {
    let id = node.type_id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    ProdType { id, fields }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, id: Option<&parser::Id>, type_id: &parser::TypeId) -> String {
        match id {
            Option::Some(id) => id.text().to_string(),
            Option::None => {
                let type_id = type_id.text();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
