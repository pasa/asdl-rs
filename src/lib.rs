mod model;
mod parser;

use tera::*;
use crate::model::*;
use failure::format_err;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

pub fn generate(asdl: &str, template: &str) -> String {
    let root: &parser::Root = &parser::parse(&asdl);
    let model = Asdl::from(root);
    let mut tera = Tera::default();
    tera.register_filter("camel", |arg, _| Ok(arg.as_str().unwrap().to_camel_case().into()));
    tera.register_filter("snake", |arg, _| Ok(arg.as_str().unwrap().to_snake_case().into()));
    tera.register_filter("mixed", |arg, _| Ok(arg.as_str().unwrap().to_mixed_case().into()));
    tera.register_filter("SCREAM", |arg, _| {
        Ok(arg.as_str().unwrap().to_shouty_snake_case().into())
    });
    tera.add_raw_template("_src", &template)
        .map_err(|e| format_err!("template parsing error: {:?}", e))
        .unwrap();
    tera.render("_src", &model)
        .map_err(|e| format_err!("template rendering error: {:?}", e))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_snapshot_matches;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Tree
            prodType = (noFileds? f, stm s1)
            ";
        let root: &parser::Root = &parser::parse(&asdl);
        assert_snapshot_matches!("simple_successful_test_syntax", root.debug_dump());
        let model = Asdl::from(root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }
}
