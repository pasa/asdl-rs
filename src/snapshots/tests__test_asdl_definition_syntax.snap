---
created: "2019-06-07T15:33:47.680034921Z"
creator: insta@0.8.1
source: src/lib.rs
expression: root.debug_dump()
---
SyntaxKind(21)@[0; 388)
  SyntaxKind(8)@[0; 3)
  SyntaxKind(22)@[3; 151)
    SyntaxKind(6)@[3; 7)
    SyntaxKind(8)@[7; 8)
    SyntaxKind(9)@[8; 9)
    SyntaxKind(8)@[9; 10)
    SyntaxKind(23)@[10; 151)
      SyntaxKind(25)@[10; 60)
        SyntaxKind(7)@[10; 13)
        SyntaxKind(26)@[13; 60)
          SyntaxKind(0)@[13; 14)
          SyntaxKind(27)@[14; 24)
            SyntaxKind(6)@[14; 24)
          SyntaxKind(5)@[24; 25)
          SyntaxKind(8)@[25; 26)
          SyntaxKind(27)@[26; 32)
            SyntaxKind(6)@[26; 31)
            SyntaxKind(4)@[31; 32)
          SyntaxKind(5)@[32; 33)
          SyntaxKind(8)@[33; 34)
          SyntaxKind(27)@[34; 45)
            SyntaxKind(6)@[34; 45)
          SyntaxKind(5)@[45; 46)
          SyntaxKind(8)@[46; 47)
          SyntaxKind(27)@[47; 59)
            SyntaxKind(6)@[47; 58)
            SyntaxKind(4)@[58; 59)
          SyntaxKind(1)@[59; 60)
      SyntaxKind(8)@[60; 89)
      SyntaxKind(2)@[89; 90)
      SyntaxKind(8)@[90; 91)
      SyntaxKind(25)@[91; 125)
        SyntaxKind(7)@[91; 98)
        SyntaxKind(26)@[98; 125)
          SyntaxKind(0)@[98; 99)
          SyntaxKind(27)@[99; 109)
            SyntaxKind(6)@[99; 109)
          SyntaxKind(5)@[109; 110)
          SyntaxKind(8)@[110; 111)
          SyntaxKind(27)@[111; 116)
            SyntaxKind(6)@[111; 116)
          SyntaxKind(5)@[116; 117)
          SyntaxKind(8)@[117; 118)
          SyntaxKind(27)@[118; 124)
            SyntaxKind(6)@[118; 123)
            SyntaxKind(4)@[123; 124)
          SyntaxKind(1)@[124; 125)
      SyntaxKind(8)@[125; 151)
  SyntaxKind(22)@[151; 214)
    SyntaxKind(6)@[151; 162)
    SyntaxKind(8)@[162; 163)
    SyntaxKind(9)@[163; 164)
    SyntaxKind(8)@[164; 165)
    SyntaxKind(23)@[165; 214)
      SyntaxKind(25)@[165; 188)
        SyntaxKind(7)@[165; 168)
        SyntaxKind(26)@[168; 188)
          SyntaxKind(0)@[168; 169)
          SyntaxKind(27)@[169; 179)
            SyntaxKind(6)@[169; 179)
          SyntaxKind(5)@[179; 180)
          SyntaxKind(8)@[180; 181)
          SyntaxKind(27)@[181; 187)
            SyntaxKind(6)@[181; 186)
            SyntaxKind(4)@[186; 187)
          SyntaxKind(1)@[187; 188)
      SyntaxKind(8)@[188; 214)
  SyntaxKind(22)@[214; 388)
    SyntaxKind(6)@[214; 219)
    SyntaxKind(8)@[219; 220)
    SyntaxKind(9)@[220; 221)
    SyntaxKind(8)@[221; 222)
    SyntaxKind(23)@[222; 388)
      SyntaxKind(25)@[222; 249)
        SyntaxKind(7)@[222; 224)
        SyntaxKind(26)@[224; 249)
          SyntaxKind(0)@[224; 225)
          SyntaxKind(27)@[225; 235)
            SyntaxKind(6)@[225; 235)
          SyntaxKind(5)@[235; 236)
          SyntaxKind(8)@[236; 237)
          SyntaxKind(27)@[237; 248)
            SyntaxKind(6)@[237; 247)
            SyntaxKind(3)@[247; 248)
          SyntaxKind(1)@[248; 249)
      SyntaxKind(8)@[249; 278)
      SyntaxKind(2)@[278; 279)
      SyntaxKind(8)@[279; 280)
      SyntaxKind(25)@[280; 311)
        SyntaxKind(7)@[280; 286)
        SyntaxKind(26)@[286; 311)
          SyntaxKind(0)@[286; 287)
          SyntaxKind(27)@[287; 297)
            SyntaxKind(6)@[287; 297)
          SyntaxKind(5)@[297; 298)
          SyntaxKind(8)@[298; 299)
          SyntaxKind(27)@[299; 310)
            SyntaxKind(6)@[299; 309)
            SyntaxKind(3)@[309; 310)
          SyntaxKind(1)@[310; 311)
      SyntaxKind(8)@[311; 340)
      SyntaxKind(2)@[340; 341)
      SyntaxKind(8)@[341; 342)
      SyntaxKind(25)@[342; 375)
        SyntaxKind(7)@[342; 350)
        SyntaxKind(26)@[350; 375)
          SyntaxKind(0)@[350; 351)
          SyntaxKind(27)@[351; 361)
            SyntaxKind(6)@[351; 361)
          SyntaxKind(5)@[361; 362)
          SyntaxKind(8)@[362; 363)
          SyntaxKind(27)@[363; 374)
            SyntaxKind(6)@[363; 373)
            SyntaxKind(3)@[373; 374)
          SyntaxKind(1)@[374; 375)
      SyntaxKind(8)@[375; 388)

